### PR TITLE
CLN: Remove Series._from_array

### LINKED ISSUE #19883: Deprecate Series._from_array ?
I open this suggestion as per @jorisvandenbossche's recommendation.  

This issue follows in the steps of #18213 and #19850.  

As it is commented in #18213, `_from_array` has a single difference with the `Series` constructor, how it handles `SparseArray`s:
```python
        # return a sparse series here
        if isinstance(arr, ABCSparseArray):
            from pandas.core.sparse.series import SparseSeries
            cls = SparseSeries
```
This process could be achieved in a similar way in `Series.__new__`; something on the lines of:
```python
def __new__( cls, *args, **kwargs ):
    # arr is mandatory, first argument or key `arr`.
    if isinstance(kwargs.get('arr', args[0]), ABCSparseArray):
        from pandas.core.sparse.series import SparseSeries
        cls = SparseSeries
    obj = object.__new__(cls)
    obj.__init__(*args, **kwargs)
    return obj
```
# What's the issue?

As @jorisvandenbossche pointed out, a change like this will result in a change of the API, as this:
```python
>>> s = pd.Series(pd.SparseArray([1, 0, 0, 2, 0]))
>>> type(s)
<class 'pandas.core.series.Series'>
```
will become this:
```python
>>> s = pd.Series.from_array(pd.SparseArray([1, 0, 0, 2, 0]))
>>> type(s)
<class 'pandas.core.sparse.series.SparseSeries'>
```
I'm not familiar with **sparse data structures**, but according to the [docs](https://pandas.pydata.org/pandas-docs/stable/sparse.html) all functionality is kept between `Series` and `SparseSeries`. Furthermore, a simple
```python
>>> s = s.to_dense()
>>> type(s)
<class 'pandas.core.series.Series'>
```
should do it to go back to `Series`.

# Why change it, then?

Currently, `Series._from_array` is called only inside two functions: `DataFrame._idx`and `DataFrame. _box_col_values`.  With the proposed change, those calls could be substituted by the default constructor.  
Being that the case, when working with `panda`'s subclassing, one would be able to declare complex `_constructor_slice` such as this:
```python
    @property
    def _constructor_sliced(self):
        def f(*args, **kwargs):
            # adapted from https://github.com/pandas-dev/pandas/issues/13208#issuecomment-326556232
            return DerivedSeries(*args, **kwargs).__finalize__(self, method='inherit')
        return f
```
, which would allow for a more complex relationship between the subclassed `DataFrame` and its sliced version, including the transfer of `metadata` according to the user's specification in `__finalize__`.

### PR DESCRIPTION
**Pull request by jaumebonet**
This PR transfers the class check of `Series._from_array` to `DataFrame._box_col_values`; removing the need for calling the `Series._from_array` and allowing for user-defined slice-inheritance of metadata.

I opted to delete the function as it was private but leave the `Series.from_array` public access working and it's FutureWarning of deprecation, as it was. To do that, I had to modify `Series.from_array` so it would keep also its original functionality until is completely deleted.

- [x] closes #19883
- [ ] passed tests.
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/frame/test_subclass.py [hunk 1]
```diff
@@ -570,3 +570,56 @@ def strech(row):
         result = df.apply(lambda x: [1, 2, 3], axis=1)
         assert not isinstance(result, tm.SubclassedDataFrame)
         tm.assert_series_equal(result, expected)
+
+    def test_frame_subclassing_and_inherit(self):
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: what is the purpose of these tests? this is another issue, yes?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -2563,8 +2562,16 @@ def _box_item_values(self, key, values):
 
     def _box_col_values(self, values, items):
         """ provide boxed values for a column """
-        return self._constructor_sliced._from_array(values, index=self.index,
-                                                    name=items, fastpath=True)
+        # This check here was previously performed in Series._from_array
+        # By doing it here there is no need for that function anymore
+        # GH#19883.
+        from pandas.core.dtypes.generic import ABCSparseArray
```
Comment (jreback) [inline pandas/core/frame.py]: make this a function like ``_get_series_result_type`` and put it near ``_get_frame_result_type`` in ``pandas/core/dtypes/concat.py``, it should just take self and return the type.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -2563,8 +2562,16 @@ def _box_item_values(self, key, values):
 
     def _box_col_values(self, values, items):
         """ provide boxed values for a column """
-        return self._constructor_sliced._from_array(values, index=self.index,
-                                                    name=items, fastpath=True)
+        # This check here was previously performed in Series._from_array
```
Comment (jreback) [inline pandas/core/frame.py]: remove all of this commentary

File: pandas/tests/frame/test_subclass.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/frame/test_subclass.py]: this should be in a new PR

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_subclass.py (hunk 1 above)
Comment (jaumebonet) [reply pandas/tests/frame/test_subclass.py]: I wrote it to check that that approach would work now. I'll take it out.

File: pandas/core/frame.py (hunk 2 above)
Comment (jaumebonet) [reply pandas/core/frame.py]: I see the idea.
Wouldn't it make more sense to be in `pandas/core/dtypes/commons.py` ?  As that file is already imported to `frame.py`?

Comment (jreback) [reply pandas/core/frame.py]: no, common is for introspection, not conversion

Comment (jaumebonet) [reply pandas/core/frame.py]: ok!

File: pandas/core/dtypes/concat.py [hunk 4]
```diff
@@ -101,6 +101,18 @@ def _get_frame_result_type(result, objs):
                                                           ABCSparseDataFrame))
 
 
+def _get_sliced_frame_result_type(data, obj):
+    """
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: add a full doc-string

File: pandas/core/dtypes/concat.py [hunk 5]
```diff
@@ -101,6 +101,18 @@ def _get_frame_result_type(result, objs):
...
+    return appropriate class of Series depending on whether
+    the data is sparse or not.
+    """
+    if is_sparse(data):
+        from pandas.core.sparse.api import SparseSeries
+        return SparseSeries
+    else:
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: just return, no else needed

File: pandas/core/dtypes/concat.py [hunk 6]
```diff
@@ -101,6 +101,18 @@ def _get_frame_result_type(result, objs):
                                                           ABCSparseDataFrame))
 
 
+def _get_sliced_frame_result_type(data, obj):
+    """
+    return appropriate class of Series depending on whether
+    the data is sparse or not.
+    """
+    if is_sparse(data):
+        from pandas.core.sparse.api import SparseSeries
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: from pandas import

File: pandas/core/frame.py [hunk 7]
```diff
@@ -2563,8 +2562,10 @@ def _box_item_values(self, key, values):
 
     def _box_col_values(self, values, items):
         """ provide boxed values for a column """
-        return self._constructor_sliced._from_array(values, index=self.index,
-                                                    name=items, fastpath=True)
+        from pandas.core.dtypes.concat import _get_sliced_frame_result_type
```
Comment (jreback) [inline pandas/core/frame.py]: this can be imported at the top

File: pandas/core/frame.py [hunk 8]
```diff
@@ -2563,8 +2562,10 @@ def _box_item_values(self, key, values):
...
+        this_constructor_sliced = _get_sliced_frame_result_type(values, self)
```
Comment (jreback) [inline pandas/core/frame.py]: call this ``klass``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/concat.py [hunk 9]
```diff
@@ -101,6 +101,27 @@ def _get_frame_result_type(result, objs):
...
 
+def _get_sliced_frame_result_type(data, obj):
+    """
+    return appropriate class of Series. When data is sparse
+    it will return a SparseSeries, otherwise it will return
+    the Series.
+
+    Parameters
+    ----------
+    data : ndarray
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: array-like

File: pandas/core/dtypes/concat.py [hunk 10]
```diff
@@ -101,6 +101,27 @@ def _get_frame_result_type(result, objs):
...
+    obj : DataFrame
+
+    Returns
+    -------
+    Series or SparseSeries
+    """
+    if is_sparse(data):
+        from pandas.core.sparse.api import SparseSeries
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: from pandas import SparseSeries

**Review by jreback — CHANGES_REQUESTED** [review]
small changes,  ping on green

**Comment by jaumebonet** [comment]
done!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks ping on green.

**Comment by jaumebonet** [comment]
sorry... I'm not sure what "ping on green" means in this context :sweat_smile:

**Comment by jreback** [comment]
it means watch this page and report when all of the checks are green.

**Comment by jreback** [comment]
thanks @jaumebonet

**Pull request merged**